pkg/notifiers/sender.go

```diff
@@ -0,0 +1,302 @@
+package notifiers
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/prometheus/common/model"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"io/ioutil"
+	"mime/multipart"
+	"net"
+	"net/http"
+	"net/smtp"
+	"net/textproto"
+	"strconv"
+	"strings"
+	"time"
+)
+
+func SendMessage(notifier *v3.Notifier, recipient string, msg *Message) error {
+	if notifier.Spec.SlackConfig != nil {
```

Comment:
Looks like we are repeating the logic of having checked `if notifier.Spec.SlackConfig` - call the `TestSlack` in 2 places - sender.go and notifier_action.go. Do you think it would make sense to introduce the method that would:

* accept notifier config
* make a decision of what driver to chose 

so both sender.go and notifier_action.go can just call this method and avoid doing if/else based checks in 2 places

Reply:
Previously the testAction Input is parsed into map so we have two different ways to do the check. Now they are consistent.